### PR TITLE
win_acl.ps1 modified to have reset func., check_mode, environment string support, and right handling corrected

### PR DESCRIPTION
**Pull request by placame**
##### SUMMARY

I've corrected the removing of rights, with post check for any remaining ACE and reporting if unsuccessful.
Added the requested reset functionality,
Added the requested support for environment variable in path,
Added check_mode support.
Also updated the doc win_acl.py

##### ISSUE TYPE

- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request

##### ADDITIONAL INFORMATION

Executed some test with below playbook:
````
---
- hosts: win-test1.adtest.celab.ibm.cloud
  gather_facts: no
  tasks:
    - name: Query ACL
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL
    - debug:
        var: GetACL.stdout_lines
    - name: Query ChildACL
      win_shell: (get-acl c:\TestFldr\SubFldr\ChildFldr).AccessToString
      register: GetChildACL
    - debug:
        var: GetChildACL.stdout_lines
    - name: Remove standard user access
      win_acl:
        path: "%SystemDrive%\\TestFldr\\SubFldr"
        user: Users
        rights: ReadAndExecute
        type: allow
        state: absent
      register: SetACL
      ignore_errors: yes
    - debug:
        var: SetACL
    - name: Query ACL1
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL1
    - debug:
        var: GetACL1.stdout_lines
    - name: add Read and Execute rights to Built-in Users
      win_acl:
        path: c:\TestFldr\SubFldr
        user: Users
        rights: ReadAndExecute
        type: allow
        state: present
    - name: Query ACL2
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL2
    - debug:
        var: GetACL2.stdout_lines
    - name: add Write rights to Built-in Users group
      win_acl:
        path: c:\TestFldr\SubFldr
        user: Users
        rights: Write
        type: allow
        state: present
    - name: Query ACL31
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL31
    - debug:
        var: GetACL31.stdout_lines
    - name: Query ChildACL31
      win_shell: (get-acl c:\TestFldr\SubFldr\ChildFldr).AccessToString
      register: GetChildACL31
    - debug:
        var: GetChildACL31.stdout_lines
    - name: revoke the previously added Write rights from Built-in Users group
      win_acl:
        path: c:\TestFldr\SubFldr
        user: Users
        rights: Write
        type: allow
        state: absent
      ignore_errors: yes
    - name: Query ACL32
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL32
    - debug:
        var: GetACL32.stdout_lines
    - name: add Read and Execute rights to TestUser
      win_acl:
        path: c:\TestFldr\SubFldr
        user: TestUser
        rights: ReadAndExecute
        type: allow
        state: present
    - name: Query ACL4
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL4
    - debug:
        var: GetACL4.stdout_lines
    - name: Query ChildACL4
      win_shell: (get-acl c:\TestFldr\SubFldr\ChildFldr).AccessToString
      register: GetChildACL4
    - debug:
        var: GetChildACL4.stdout_lines
    - name: add Modify rights to TestUser
      win_acl:
        path: c:\TestFldr\SubFldr
        user: TestUser
        rights: Modify
        type: allow
        state: present
    - name: Query ACL41
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL41
    - debug:
        var: GetACL41.stdout_lines
    - name: Query ChildACL41
      win_shell: (get-acl c:\TestFldr\SubFldr\ChildFldr).AccessToString
      register: GetChildACL41
    - debug:
        var: GetChildACL41.stdout_lines
    - name: revoke Modify rights from TestUser
      win_acl:
        path: c:\TestFldr\SubFldr
        user: TestUser
        rights: Modify
        type: allow
        state: absent
    - name: Query ACL42
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL42
    - debug:
        var: GetACL42.stdout_lines
    - name: restore the expected state
      win_acl:
        path: c:\TestFldr\SubFldr
        user: TestUser
        rights: ReadAndExecute
        type: allow
        state: present
    - name: Query ACL43
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL43
    - debug:
        var: GetACL43.stdout_lines
    - name: add Write rights to TestUser
      win_acl:
        path: c:\TestFldr\SubFldr
        user: TestUser
        rights: Write
        type: allow
        state: present
    - name: Query ACL44
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL44
    - debug:
        var: GetACL44.stdout_lines
    - name: revoke Write rights from TestUser
      win_acl:
        path: c:\TestFldr\SubFldr
        user: TestUser
        rights: Write
        type: allow
        state: absent
    - name: Query ACL45
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL45
    - debug:
        var: GetACL45.stdout_lines
    - name: Reset ACL
      win_acl:
        path: c:\TestFldr\SubFldr
        state: reset
      register: ReSetACL
      ignore_errors: yes
    - debug:
        var: ReSetACL
    - name: Query ACL5
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL5
    - debug:
        var: GetACL5.stdout_lines
    - name: Reset ACL on root (M:\)
      win_acl:
        path: m:\
        state: reset
      register: ReSetACL
      ignore_errors: yes
    - debug:
        var: ReSetACL
    - name: Query ACL6
      win_shell: (get-acl c:\TestFldr\SubFldr).AccessToString
      register: GetACL6
    - debug:
        var: GetACL6.stdout_lines

````

The result:
```paste below

PLAY [win-test1.adtest.celab.ibm.cloud] ****************************************

TASK [Query ACL] ***************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL.stdout_lines": [
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [Query ChildACL] **********************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetChildACL.stdout_lines": [
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [Remove standard user access] *********************************************
fatal: [win-test1.adtest.celab.ibm.cloud]: FAILED! => {"changed": false, "msg": "Users still has CreateFiles, AppendData, ReadAndExecute, Synchronize rights! Actual rights: Allow: ReadAndExecute, Synchronize: Inherited, ContainerInherit, ObjectInherit; Allow: AppendData: Inherited, ContainerInherit; Allow: CreateFiles: Inherited, ContainerInherit; ", "stderr": "Users still has CreateFiles, AppendData, ReadAndExecute, Synchronize rights!", "stderr_lines": ["Users still has CreateFiles, AppendData, ReadAndExecute, Synchronize rights!"]}
...ignoring

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "SetACL": {
        "changed": false,
        "failed": true,
        "msg": "Users still has CreateFiles, AppendData, ReadAndExecute, Synchronize rights! Actual rights: Allow: ReadAndExecute, Synchronize: Inherited, ContainerInherit, ObjectInherit; Allow: AppendData: Inherited, ContainerInherit; Allow: CreateFiles: Inherited, ContainerInherit; ",
        "stderr": "Users still has CreateFiles, AppendData, ReadAndExecute, Synchronize rights!",
        "stderr_lines": [
            "Users still has CreateFiles, AppendData, ReadAndExecute, Synchronize rights!"
        ]
    }
}

TASK [Query ACL1] **************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL1.stdout_lines": [
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [add Read and Execute rights to Built-in Users] ***************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [Query ACL2] **************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL2.stdout_lines": [
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [add Write rights to Built-in Users group] ********************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [Query ACL31] *************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL31.stdout_lines": [
        "BUILTIN\\Users Allow  Write, ReadAndExecute, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [Query ChildACL31] ********************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetChildACL31.stdout_lines": [
        "BUILTIN\\Users Allow  Write, ReadAndExecute, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [revoke the previously added Write rights from Built-in Users group] ******
fatal: [win-test1.adtest.celab.ibm.cloud]: FAILED! => {"changed": true, "msg": "Users still has CreateFiles, AppendData, ReadAndExecute, Synchronize rights! New rights: Allow: ReadAndExecute, Synchronize:  ContainerInherit, ObjectInherit; Allow: ReadAndExecute, Synchronize: Inherited, ContainerInherit, ObjectInherit; Allow: AppendData: Inherited, ContainerInherit; Allow: CreateFiles: Inherited, ContainerInherit; ", "stderr": "Users still has CreateFiles, AppendData, ReadAndExecute, Synchronize rights!", "stderr_lines": ["Users still has CreateFiles, AppendData, ReadAndExecute, Synchronize rights!"]}
...ignoring

TASK [Query ACL32] *************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL32.stdout_lines": [
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [add Read and Execute rights to TestUser] *********************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [Query ACL4] **************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL4.stdout_lines": [
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "WIN-TEST1\\TestUser Allow  ReadAndExecute, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [Query ChildACL4] *********************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetChildACL4.stdout_lines": [
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "WIN-TEST1\\TestUser Allow  ReadAndExecute, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [add Modify rights to TestUser] *******************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [Query ACL41] *************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL41.stdout_lines": [
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "WIN-TEST1\\TestUser Allow  Modify, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [Query ChildACL41] ********************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetChildACL41.stdout_lines": [
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "WIN-TEST1\\TestUser Allow  Modify, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [revoke Modify rights from TestUser] **************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [Query ACL42] *************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL42.stdout_lines": [
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [restore the expected state] **********************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [Query ACL43] *************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL43.stdout_lines": [
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "WIN-TEST1\\TestUser Allow  ReadAndExecute, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [add Write rights to TestUser] ********************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [Query ACL44] *************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL44.stdout_lines": [
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "WIN-TEST1\\TestUser Allow  Write, ReadAndExecute, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [revoke Write rights from TestUser] ***************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [Query ACL45] *************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL45.stdout_lines": [
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "WIN-TEST1\\TestUser Allow  ReadAndExecute, Synchronize",
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [Reset ACL] ***************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "ReSetACL": {
        "changed": true,
        "failed": false,
        "msg": ""
    }
}

TASK [Query ACL5] **************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL5.stdout_lines": [
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

TASK [Reset ACL on root (M:\)] *************************************************
fatal: [win-test1.adtest.celab.ibm.cloud]: FAILED! => {"changed": false, "msg": "m:\\ is a root folder! Cannot reset ACL!"}
...ignoring

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "ReSetACL": {
        "changed": false,
        "failed": true,
        "msg": "m:\\ is a root folder! Cannot reset ACL!"
    }
}

TASK [Query ACL6] **************************************************************
changed: [win-test1.adtest.celab.ibm.cloud]

TASK [debug] *******************************************************************
ok: [win-test1.adtest.celab.ibm.cloud] => {
    "GetACL6.stdout_lines": [
        "NT AUTHORITY\\SYSTEM Allow  FullControl",
        "BUILTIN\\Administrators Allow  FullControl",
        "BUILTIN\\Users Allow  ReadAndExecute, Synchronize",
        "BUILTIN\\Users Allow  AppendData",
        "BUILTIN\\Users Allow  CreateFiles",
        "CREATOR OWNER Allow  268435456"
    ]
}

PLAY RECAP *********************************************************************
win-test1.adtest.celab.ibm.cloud : ok=49   changed=27   unreachable=0    failed=0    skipped=0    rescued=0    ignored=3   


```

### CODE REVIEW

**Review by jborean93 — CHANGES_REQUESTED** [review]
I've left a few comments but as the PR stands now it will require a few things before we could look at merging it in. This module isn't the easiest to understand and unfortunately isn't tested too well which makes it hard to review these changes.

What I recommend is to split up each of the features you want to introduce into smaller PRs. Creating a separate PR for the each of the following will make it a lot easier for us to review and eventually merge in the changes:

* Doc updates
* Support for env vars in the path
* Check mode
* `state: reset` functionality

By splitting up the code we can merge in your changes piece by piece and potentially a lot quicker than trying to make a massive change all at once. It also makes the history a lot easier to understand by seeing each change as a separate layer rather than mixed together.

The last thing I should mention is that any new functionality in the code should have some integration tests to test the new functionality. This is a hard requirement and we won't be merging any PRs related to code changes without tests.

If you need help with trying to add integration tests I'm more than happy to help where I can. For now you can find our current tests at https://github.com/ansible-collections/ansible.windows/tree/main/tests/integration/targets/win_acl.

File: plugins/modules/win_acl.ps1 [hunk 1]
```diff
@@ -78,18 +85,13 @@ Function SetPrivilegeTokens() {
...
-$type = Get-AnsibleParam -obj $params -name "type" -type "str" -failifempty $true -validateset "allow","deny"
-$state = Get-AnsibleParam -obj $params -name "state" -type "str" -default "present" -validateset "absent","present"
-
-$inherit = Get-AnsibleParam -obj $params -name "inherit" -type "str"
-$propagation = Get-AnsibleParam -obj $params -name "propagation" -type "str" -default "None" -validateset "InheritOnly","None","NoPropagateInherit"
-
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -default $false
+# Get the path parameter with expanded environment variables.
+$path=Get-AnsibleParam -obj $params -name "path" -type "str" -failifempty $true
+$path = (New-Object -ComObject Wscript.Shell).ExpandEnvironmentStrings($path)
```
[the remark this answers is not in the document]
Comment (placame) [reply plugins/modules/win_acl.ps1:94]: Will this `-type "path"` accept registry paths as well?

File: plugins/modules/win_acl.ps1 [hunk 2]
```diff
@@ -3,6 +3,13 @@
 # Copyright: (c) 2015, Phil Schwartz <schwartzmx@gmail.com>
 # Copyright: (c) 2015, Trond Hindenes
 # Copyright: (c) 2015, Hans-Joachim Kliemeck <git@kliemeck.de>
+# Revorked the logic to more exact handle of rights and proper return values,
```
[the remark this answers is not in the document]
Comment (placame) [reply plugins/modules/win_acl.ps1]: I suppose my changes are substantial, as I've completely re-wrote the full logic to be able to gracefully fail when an inherited rights needed to be revoked. Also the 'reset' functionality is important thing. So added myself to the list, and removed everything else.

File: plugins/modules/win_acl.ps1 [hunk 3]
```diff
@@ -78,18 +85,13 @@ Function SetPrivilegeTokens() {
 
 $result = @{
     changed = $false
+	msg = ""
 }
 
-$path = Get-AnsibleParam -obj $params -name "path" -type "str" -failifempty $true
```
[the remark this answers is not in the document]
Comment (placame) [reply plugins/modules/win_acl.ps1:83]: I've restructured the script so the parameter usage is now kept together as much as possible.

File: plugins/modules/win_acl.ps1 [hunk 4]
```diff
@@ -104,8 +106,82 @@ if ($path_qualifier -eq "HKCC:" -and (-not (Test-Path -LiteralPath HKCC:\))) {
...
+         }
+      if($changed){
+	     If ($path_item.PSProvider.Name -eq "Registry") {
+                Set-ACL -LiteralPath $path -AclObject $objACL
+            } else {
+                (Get-Item -LiteralPath $path).SetAccessControl($objACL)
+            }
+         }
+      # When check_mode, the saved ACL needs to be written back to restore the original config.
+	  if($check_mode){
```
[the remark this answers is not in the document]
Comment (placame) [reply plugins/modules/win_acl.ps1]: Re-worked this check_mode logic, now it does not change anything.

File: plugins/modules/win_acl.ps1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (placame) [reply plugins/modules/win_acl.ps1]: I'd rather keep the type as string and perform later the necessary expansions to have more control over the process.

**Comment by placame** [comment]
@jborean93 
I'd like to ask your help regarding the integration tests.
The current integration tests are not intended to show results of the new functionalities, also not showing the result of code correction regarding inherited rights.
Should I some way copy the current tests, and modify it to show the missing parts, or add new parts to the existing tests?
Can you help me with a way of executing the current integration tests on a local environment?

**Comment by jborean93** [comment]
> Should I some way copy the current tests, and modify it to show the missing parts, or add new parts to the existing tests?

You should leave the existing tests as they are today and just add more tests for your use case. Usually what we try to encourage for tests is to do the following:

```yaml
- name: run test in check mode
  win_acl:
    ...
  check_mode: yes
  register: scenario_check

-name: get result of run test in check mode
  win_shell: ...
  register: scenario_actual_check

- name: assert run test in check mode
  assert:
    that:
    # Verifies in check mode that it reported a change but didn't actually change the underlying resource
    - scenario_check is changed
    - scenario_actual_check.stdout == ...

- name: run test
  win_acl:
    ...
  register: scenario

-name: get result of run test
  win_shell: ...
  register: scenario_actual

- name: assert run test
  assert:
    that:
    # Verifies it reported a change and changed the resource to what we expected
    - scenario is changed
    - scenario_actual.stdout == ...

- name: run test idempotent
  win_acl:
    ...
  register: scenario_again

- name: assert run test idempotent
  assert:
    that:
    # On a 2nd run the task should not report a change
    - not scenario_again is changed
```

Usually I only have a check mode test for a few of them, or just one scenario that I know exercises the check mode code itself. The `win_shell` task is meant to get the actual state of the resource. Usually I just get the object output it as json then use `scenario_result.stdout | from_json` to be able to parse it in Ansible. I would recommend having a look at [win_certificate_store](https://github.com/ansible-collections/ansible.windows/tree/main/tests/integration/targets/win_certificate_store) and [win_dns_client](https://github.com/ansible-collections/ansible.windows/tree/main/tests/integration/targets/win_dns_client) for an example of some tests.

> Can you help me with a way of executing the current integration tests on a local environment?

Yep, all you need to do is place a file at `tests/integration/inventory.winrm` with the content

```ini
[windows]
my-host

[windows:vars]
ansible_user=username
ansible_password=password
ansible_connection=winrm
...
```

You will need to set whatever vars Ansible needs to connect to the host. You don't need to use the `winrm` connection plugin, just any one that Ansible can use to connect to your Windows host. Basically you should be able to do `ansible -i tests/integration/inventory.winrm windows -m win_ping` with that inventory file.

Let me know if you have any more questions.

**Comment by placame** [comment]
@jborean93 
I've created some integration test. 
Find the test in this archive: 
[win_acl_integration_test.zip](https://github.com/ansible-collections/ansible.windows/files/5615626/win_acl_integration_test.zip)
TestResult.txt contains the output of: `ansible-playbook -i hosts ./win_acl_tests.yml >./TestResult.txt`
Let me know if any other thing, or more test would be required.
Or the tests should be executed with `-vvvv` option

**Comment by placame** [comment]
@jborean93 
Any news or progress? 

I cannot split it into smaller changes as the solution for the new requirements are based on the changes in the main logic correcting errors in the original method.

**Comment by placame** [comment]
@jborean93 ??

**Comment by placame** [comment]
Hi @jborean93 ,
Ok, so here is the misunderstandings. The main purpose of my development was to correct the **errors** in the original code / **logic**. the added features are only secondary things. But correcting the original faulty code and logic makes it absolutely necessary to change almost the whole code, at least the part which hosted the faulty logic. You cannot except correcting a faulty logic with very similar code.

**Comment by jborean93** [comment]
That's the thing you are correcting some behaviour but adding new features. They are in no way related, you can do 1 without the other. To implement reset behaviour you don't need check mode, environment support doesn't need reset or check mode and vice versa. Putting it all together just makes it a whole lot harder for myself to review.

I know it's not ideal but these are supported modules so we need to treat it with a higher standard and higher scrutiny when compared to `community.windows`. This means a proper reivew which right now just isn't feasible with everything bundled together.

**Comment by placame** [comment]
Ok, I'll open new pull requests for the stages. first the correction of the faulty logic.

**Pull request closed without merging**